lib/Optimizer/Transforms/DependencyAnalysis.cpp

```diff
@@ -0,0 +1,3148 @@
+          freevars.remove(shadowNode);
+          delete shadowNode;
+        }
+      } else if (dependency->isLeaf() && dependency->isQuantumOp()) {
+        // The dependency is a block argument, and therefore reflects a
+        // dependency for this `if` First, find the relevant argument
+        ArgDependencyNode *arg =
+            static_cast<ArgDependencyNode *>(dependency.node);
+        auto num = arg->getArgNumber();
+        // Then, get the dependency from this `if` for the relevant argument,
+        // this will be the new dependency for `then_op`
+        auto newDep = dependencies[num + 1];
+        newDep->successors.remove(this);
+        newDep->successors.insert(then_op);
+        newDeps.push_back(newDep);
+        arg->successors.remove(then_op);
+
+        // Replace the dependency with the relevant result from the lifted node
+        dependencies[num + 1] =
+            DependencyEdge{then_op, then_op->getResultForDependency(i)};
+
+        // Remove then_op from the route for then_qid inside the block
+        then_op->eraseEdgeForQID(dependency.qid.value());
+      } else if (!dependency->isQuantumOp()) {
+        newDeps.push_back(dependency);
+      } else {
+        assert(
+            false &&
+            "Trying to lift a quantum operation before dependency was lifted");
+      }
+    }
+
+    else_op->erase();
+    delete else_op;
+
+    // Patch successors
+    then_op->successors.insert(this);
+    then_op->dependencies = newDeps;
+  }
+
+  /// Lifts a qubit allocated in the then/else blocks to be allocated
+  /// in the graph \p parent containing this `if`. Adds a new linear argument,
+  /// depending on the lifted alloc, and corresponding result, depended on by
+  /// the lifted dealloc, to this `if`. If the phyiscal qubit is present in an
+  /// inner block, the inner alloc/de-alloc pair will be removed. Both blocks
+  /// will have the new argument/terminator dependency added, so the wire flows
+  /// through the block properly even if it is not used.
+  ///
+  /// Currently, all allocations are lifted from `if`s, so they can be combined
+  /// in the parent context. This is legal, as `if`s are treated as "solid
+  /// barriers" in the parent graph, so allocating before/after the `if` is
+  /// equivalent to allocating within the `if`. This effectively undoes
+  /// contractAllocsPass after having allocs be contracted for performing the
+  /// analysis is no longer helpful.
+  ///
+  /// This is not necessary to always perform, instead, one could only lift
+  /// allocs when lifting operations on those allocs, but it is very difficult
+  /// to do that safely.
+  ///
+  /// Assumes (and checks) that \p qubit is allocated in either/both the
+  /// then/else blocks.
+  ///
+  /// If the qubit is present in both child blocks, then the extra
+  /// alloc/de-alloc pair is cleaned up here.
+  void liftAlloc(PhysicalQID qubit, DependencyGraph *parent) {
```

Comment:
Is this only looking for common prefixes across the arms of `if`s? Or is there someplace that is doing tail fusion and sinking common tail suffixes?

Reply:
`tryLiftingBefore` looks for common prefixes (and `liftOpBefore` does the lifting), and `tryLiftingAfter` looks for common "postfixes" (and `liftOpAfter` does the lifting). `liftAlloc` lifts all physical allocations to the parent scope.

Reply:
I see. Thanks for clarifying.